test/test_vmap.py

```diff
@@ -1150,5 +1159,100 @@ def test_no_random_op_support(self):
+    #
+    # output_process_fn: a function that maps the outputs to the part
+    #       that should be differentiated.
+    # batch_size: the batch dim size for the batched grad
+    def _batched_grad_test(self, op, args, kwargs, output_process_fn=lambda x: x, batch_size=3):
+        outputs = op(*args, **kwargs)
+        outputs = differentiable(output_process_fn(outputs))
+        batched_vectors = tuple(construct_v(out, batch_size) for out in outputs)
+
+        def vector_jacobian_product(*vectors):
+            return torch.autograd.grad(outputs, differentiable(args), vectors,
+                                       retain_graph=True)
+        self._vmap_test(vector_jacobian_product, batched_vectors,
+                        check_propagates_grad=False)
+
+    # Tests batched second grad computation of outputs = op(*args, **kwargs).
+    # by comparing it to a sequential map+stack fallback.
+    #
+    # output_process_fn: a function that maps the outputs to the part
+    #       that should be differentiated.
+    # batch_size: the batch dim size for the batched grad
+    #
+    # NB: we only test computing batched gradients are in the second gradient
+    # computation. One specific use case that does this is computing the hessian
+    # matrix of a scalar-valued function; this is useful in Bayesian Logistic
+    # Regression.
+    # It might be useful to have a test that computes batched first gradients and
+    # then uses those to compute batched second gradients in the future.
+    def _batched_grad_grad_test(self, op, args, kwargs, output_process_fn=lambda x: x, batch_size=3):
+        outputs = op(*args, **kwargs)
+        outputs = differentiable(output_process_fn(outputs))
+        ones = tuple(torch.ones_like(out) for out in outputs)
+        # Same thing as summing together all of the outputs and calling .backward()
+        first_grads = torch.autograd.grad(outputs, differentiable(args), ones,
+                                          create_graph=True)
+        first_grads = differentiable(first_grads)
+        if len(first_grads) == 0:
+            return  # None of the grads actually depend on the input
+
+        batched_vectors = tuple(construct_v(grad, batch_size) for grad in first_grads)
+
+        def vector_hessian_product(*vectors):
+            outputs = torch.autograd.grad(first_grads, differentiable(args), vectors,
+                                          retain_graph=True, allow_unused=True)
+            outputs = tuple(out for out in outputs if out is not None)
+            assert len(outputs) > 0
+            return outputs
+
+        self._vmap_test(vector_hessian_product, batched_vectors,
+                        check_propagates_grad=False)
+
+    def test_sigmoid(self, device):
+        # Maybe we can make the "check that the slow fallback was not invoked"
+        # into a context manager, because it's used a lot. I'll leave that for
+        # future work.
```

Comment:
Yes, the Not regex here looks very brittle; you won't notice it's out of date if you change the message

Reply:
Yup, I made it more structured in another PR (https://github.com/pytorch/pytorch/pull/43731)